src/IssueViz.UnitTests/Helpers/RunOnUIThreadTests.cs

```diff
@@ -0,0 +1,172 @@
+using Moq;
+using SonarLint.VisualStudio.Integration.UnitTests;
+using SonarLint.VisualStudio.IssueVisualization.Helpers;
+
+namespace SonarLint.VisualStudio.IssueVisualization.UnitTests.Helpers
+{
+    [TestClass]
+    public class RunOnUIThreadTests
+    {
+        // These tests are slightly hacky in that they use "ThreadHelper.SetCurrentThreadAsUIThread()"
+        // to fool the VS thread helper into thinking the current thread is the UI thread.
+        // 
+        // However, these tests are quite limited as we can only test the cases where the
+        // caller is already on the main thread - we're not hacking/mocking/faking enough 
+        // of the VS setup for a call to "ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync()"
+        // to succeed.
+
+        [TestMethod]
+        public void Run_StartsOnUIThread_ExecuteOnUIThread()
+        {
+            ThreadHelper.SetCurrentThreadAsUIThread();
+
+            var initialThread = Thread.CurrentThread.ManagedThreadId;
+            int executionThreadId = -1;
+
+            // Act
+            RunOnUIThread.Run(() => executionThreadId = Thread.CurrentThread.ManagedThreadId);
+
+            executionThreadId.Should().Be(initialThread);
+            Thread.CurrentThread.ManagedThreadId.Should().Be(initialThread); // should still be on the "UI" thread
+        }
+
+        [TestMethod]
+        public async Task RunAsync_StartsOnUIThread_ExecuteOnUIThread()
+        {
+            ThreadHelper.SetCurrentThreadAsUIThread();
+
+            var initialThread = Thread.CurrentThread.ManagedThreadId;
+            int executionThreadId = -1;
+
+            // Act
+            await RunOnUIThread.RunAsync(() => executionThreadId = Thread.CurrentThread.ManagedThreadId);
+
+            executionThreadId.Should().Be(initialThread);
+            Thread.CurrentThread.ManagedThreadId.Should().Be(initialThread); // should still be on the "UI" thread
+        }
+    }
+
+    [TestClass]
+    public class RunOnUIThreadTests_Hacky
+    {
+        // These tests are quite hacky in that they use "ThreadHelper.SetCurrentThreadAsUIThread()",
+        // and also use reflection to set private data fields in the VS ThreadHelper hierarchy.
+        // As such, they will break if the underlying VS implementation changes.
+        // to fool the VS thread helper into thinking the current thread is the UI thread.
+        // 
+        // However, we are now hacking/mocking/faking enough of the VS setup that a call to 
+        // "ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync()" won't fail.
+        // It won't actually cause a thread transition, but we can use the fact that our
+        // scheduler mock was called to detect whether a thread transition was requested.
+
+        private Mock<IVsTaskSchedulerService2> schedulerMock;
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            ThreadHelper.SetCurrentThreadAsUIThread();
```

Comment:
Do we need this for all the tests?

For tests like `Run_StartsOnBackgroundThread_ExecuteOnUIThread` we can avoid starting a new task if we don't do this.

Reply:
Unfortunately not. Each test passes if you run it individually, but not if you run them as a group. It's because the 'ThreadHelper' test code is setting a static, any we currently have no way of resetting the effect. So to maintain isolation it's better if each test explicitly sets what is considered the UI thread for that test.

I can improve the test isolation slightly for the really hacky tests by making them reset the field they are setting, so I'll do that. We might be able to do something similar for the test ThreadHelper, but I won`t do that now.

Reply:
Ah, I see. 
Maybe we should change ThreadHelper to return a disposable scope, like AssertIgnoreScope. I'll open a ticket.

Reply:
FYI I've added a disposable scope that deals with the `cachedServiceInstance` field, but I haven't tried to unset whatever is set by the ThreadHelper call.